File: solutions/go/roman-numerals/1/roman_numerals.go
package romannumerals

import "errors"

type numeral struct {
	decimal int
	roman   string
}

var decimalToRoman = [13]numeral{
	{1000, "M"},
	{900, "CM"},
	{500, "D"},
	{400, "CD"},
	{100, "C"},
	{90, "XC"},
	{50, "L"},
	{40, "XL"},
	{10, "X"},
	{9, "IX"},
	{5, "V"},
	{4, "IV"},
	{1, "I"},
}

func ToRomanNumeral(input int) (string, error) {
	if input >= 4000 || input <= 0 {
		return "", errors.New("input cannot be represented in roman numerals")
	}
	output := ""
	for _, numeral := range decimalToRoman {
		for input >= numeral.decimal {
			output += numeral.roman
			input -= numeral.decimal
		}
	}
	return output, nil
}

File: solutions/kotlin/roman-numerals/1/src/main/kotlin/RomanNumerals.kt
object RomanNumerals {

    private val romanToArabic = mapOf(
        1000 to "M",
        900 to "CM",
        500 to "D",
        400 to "CD",
        100 to "C",
        90 to "XC",
        50 to "L",
        40 to "XL",
        10 to "X",
        9 to "IX",
        5 to "V",
        4 to "IV",
        1 to "I"
    )

    fun value(n: Int): String {
        var input = n
        var output = ""
        for ((decimal, arabic) in romanToArabic) {
            while (input >= decimal) {
                output += arabic
                input -= decimal
            }
        }
        return output
    }
}
